java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -89,11 +94,28 @@ abstract class AbstractArcHost(
         initialParticles.toList().associateByTo(particleConstructors, { it.first }, { it.second })
     }
 
+    private val cacheMutex = Mutex()
+    private val runningMutex = Mutex()
+
+    private suspend fun putContextCache(id: String, context: ArcHostContext) = cacheMutex.withLock {
+        contextCache[id] = context
+    }
+
+    private suspend fun clearContextCache() = cacheMutex.withLock {
+        contextCache.clear()
+    }
+
+    private suspend fun getContextCache(arcId: String) = cacheMutex.withLock {
+        contextCache[arcId]
+    }
+
     /**
      * Determines if [arcId] is currently running. It's state must be [ArcState.Running] and
      * it must be memory resident (not serialized and dormant).
      */
-    protected fun isRunning(arcId: String) = runningArcs[arcId]?.arcState == ArcState.Running
+    protected fun isRunning(arcId: String) = runBlocking {
+        runningMutex.withLock { runningArcs[arcId]?.arcState == ArcState.Running }
+    }
```

Comment:
Locking a mutex inside runBlocking seems a bit like just using java's synchronized keyword, but with extra steps (and is possibly slower? Not sure)

Do you know if synchronized works for Kotlin multiplatform? Do you think that would be easier to use here? It's just a bit annoying scattering `suspend` everywhere.

Reply:
Synchronized works, but there's a lint/compiler warning if you try to have a suspend function that uses synchronized. :(

I didn't want to add any runBlocking, I could remove it by making these suspend functions, but then I'd need to go find the G3 overriders, but I though it would take too long to land given the branch cut coming tonite. 

Another possibility is copy-on-write, making an atomic(map), get rid of mutableMap, and force all updates to clone the map.

Are you ok with me doing that in a follow up?

Reply:
Ah I see. I didn't realise that. I just read this article which was quite informative: https://blog.danlew.net/2020/01/28/coroutines-and-java-synchronization-dont-mix/

I think runBlocking + withLock is probably the next best alternative then. I don't think you should make these all suspend functions, and the copy-on-write sounds like it might be more trouble than it's worth.

It's a shame we can't just use Java's ConcurrentHashMap :-(

Reply:
Yea, I did that once before in core (used ConcurrentHashMap) and Jason bopped me. :) Seems someone needs to port it to Kotlin. :)

I dunno if you're ever seen Cliff Click's GoogleTalk on concurrent hash map, but it's quite illuminating: https://www.youtube.com/watch?v=HJ-719EGIts